Mono.Debugging.Soft/SoftDebuggerSession.cs

```diff
@@ -2640,58 +2640,62 @@ bool LoadPdbFile (string pdbFileName)
-
-					var document = pdb.GetDocument (documentHandle);
-					info.Hash = pdb.GetBlobBytes (document.Hash);
-					info.HashAlgorithem = pdb.GetGuid (document.HashAlgorithm);
-					info.FileID = 0;
-					info.FullFilePath = pdb.GetString (document.Name);
-
-					fileToSourceFileInfos [info.FullFilePath] = new List<SourceFileDebugInfo> ();
-
-					if (!fileToSourceFileInfos.ContainsKey (Path.GetFileName (info.FullFilePath)))
-						fileToSourceFileInfos [Path.GetFileName (info.FullFilePath)] = new List<SourceFileDebugInfo> ();
-
-					fileToSourceFileInfos [info.FullFilePath].Add (info);
-					fileToSourceFileInfos [Path.GetFileName (info.FullFilePath)].Add (info);
-				}
-			}
-
-			return true;
-		}
-
+			var fileToSourceFileInfos = new Dictionary<string, List<SourceFileDebugInfo>>(StringComparer.InvariantCultureIgnoreCase);
+			sourceFilesDebugInfo[pdbFileName] = fileToSourceFileInfos;
+
+
+			using (var module = Cecil.ModuleDefinition.ReadModule(Path.ChangeExtension(pdbFileName, ".dll")))
+			{
+				var symbolReaderProvider = new Cecil.Cil.DefaultSymbolReaderProvider(false);
+				module.ReadSymbols(symbolReaderProvider.GetSymbolReader(module, pdbFileName));
+
+				var methodMapping = new Dictionary<Document, List<SequencePoint[]>>();
+				foreach (var type in module.GetTypes())
+					LoadPdbType(type, fileToSourceFileInfos);
+			}
+
+			return true;
+		}
+
+		private void LoadPdbType(Cecil.TypeDefinition type, Dictionary<string, List<SourceFileDebugInfo>> fileToSourceFileInfos)
+		{
+			foreach (var method in type.Methods)
+			{
+				var documents = method.DebugInformation.SequencePoints.GroupBy(t => t.Document, t => t);
+
+				foreach (var document in documents)
+				{
+					AddSequencePointsToFileSource(fileToSourceFileInfos, document.Key, document, document.Key.Url);
+					AddSequencePointsToFileSource(fileToSourceFileInfos, document.Key, document, Path.GetFileName(document.Key.Url));
+				}
+			}
+		}
+
+		private static void AddSequencePointsToFileSource(Dictionary<string, List<SourceFileDebugInfo>> fileToSourceFileInfos, Document document, IEnumerable<SequencePoint> sequencePoints, string name)
+		{
+			if (!fileToSourceFileInfos.ContainsKey(name))
+			{
+				List<SequencePoint[]> list = new List<SequencePoint[]> { sequencePoints.ToArray() };
+				SourceFileDebugInfo info = new SourceFileDebugInfo(list);
+
+				info.Hash = document.Hash;
+				info.FileID = 0;
+				info.FullFilePath = document.Url;
+
+				fileToSourceFileInfos.Add(name, new List<SourceFileDebugInfo> { info });
+			}
+			else
+			{
+				fileToSourceFileInfos[name][0].PdbMethods.Add(sequencePoints.ToArray());
+			}
+		}
+
 		bool LoadMdbFile (string mdbFileName)
 		{
 			if (!File.Exists (mdbFileName))
 				return false;
 
-			var fileToSourceFileInfos = new Dictionary<string, List<SourceFileDebugInfo>> ();
+			var fileToSourceFileInfos = new Dictionary<string, List<SourceFileDebugInfo>> (StringComparer.InvariantCultureIgnoreCase);

```

Comment:
I think this is more an Id than a path. For starters, there's two copies of each (one with full path and one with just the file name), so I'm not sure how that will work :) Somebody else may know better.